File: qyao/121-best-time-to-buy-and-sell-stock/Solution.java
class Solution {
  /*
   * Going from left to right while keeping track of 1. smallest element so far 2. highest (max - min) so far
   */
  public int maxProfit(int[] prices) {
    if (prices.length == 0) {
      return 0;
    }
    int smallestSoFar = prices[0];
    int highestProfitSoFar = 0;
    for (int price : prices) {
      smallestSoFar = Math.min(smallestSoFar, price);
      highestProfitSoFar = Math.max(highestProfitSoFar, price - smallestSoFar);
    }
    return highestProfitSoFar;
  }
}

File: qyao/121-best-time-to-buy-and-sell-stock/Solution.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // doing one-pass: keeping track of two variables, min-price and max-profit
        // if the next node is lower than this one, update min-price
        // if it is greater, then see if the profit exceeds the maximum profit
        if (prices.size() == 0) {
            return 0;
        }
        
        int min_price = prices[0];
        int max_profit = 0;
        
        for (int i = 1; i < prices.size(); i++) {
            if (prices[i] < min_price) {
                min_price = prices[i];
            }
            else {
                max_profit = max(max_profit, prices[i] - min_price);
            }
        }
        
        return max_profit;
    }
};
